drivers/usb/device/usb_dc_nrfx.c

```
@@ -515,6 +517,136 @@ static void usbd_enable_endpoints(struct nrf_usbd_ctx *ctx)
 	}
 }
 
+/**
+ * @brief Reset endpoint state.
+ *
+ * Resets the internal logic state for a given endpoint.
+ *
+ * @param[in]  ep_cts   Endpoint structure control block
+ */
+static inline void ep_ctx_reset(struct nrf_usbd_ep_ctx *ep_ctx)
+{
+	ep_ctx->buf.data = ep_ctx->buf.block.data;
+	ep_ctx->buf.curr = ep_ctx->buf.data;
+	ep_ctx->buf.len  = 0;
+
+	ep_ctx->read_complete = true;
+	ep_ctx->read_pending = false;
+	ep_ctx->write_in_progress = false;
+}
+
+/**
+ * @brief Initialize all endpoint structores.
+ *
+ * Endpoint buffers are allocated during the first call of this function.
+ * This function may also be called again on every USB reset event
+ * to reinitialize the state of all endpoints.
+ */
+static void eps_ctx_init(void)
+{
+	struct nrf_usbd_ep_ctx *ep_ctx;
+	int ret;
+	u32_t i;
+
+	for (i = 0; i < CFG_EPIN_CNT; i++) {
+		ep_ctx = in_endpoint_ctx(i);
+		__ASSERT_NO_MSG(ep_ctx);
+
+		if (!ep_ctx->buf.block.data) {
+			ret = k_mem_pool_alloc(&ep_buf_pool, &ep_ctx->buf.block,
+					       MAX_EP_BUF_SZ, K_NO_WAIT);
+			if (ret < 0) {
+				LOG_ERR("EP buffer alloc failed for EPIN%d", i);
+				__ASSERT_NO_MSG(0);
```

maybe this function should return an error?